readonly_bs.go

```diff
@@ -0,0 +1,207 @@
+
+// ShardSelectorF helps select a shard to fetch a cid from if the given cid is present in multiple shards.
+type ShardSelectorF func(c cid.Cid, shards []shard.Key) (shard.Key, error)
+
+type accessorWithBlockstore struct {
+	sa *ShardAccessor
+	bs ReadBlockstore
+}
+
+// AllShardsReadBlockstore is a read only blockstore over all cids across all shards in the dagstore.
+type AllShardsReadBlockstore struct {
+	d            *DAGStore
+	shardSelectF ShardSelectorF
+
+	// caches the blockstore for a given shard for shard read affinity i.e. further reads will likely be from the same shard.
+	// shard key -> read only blockstore
+	blockstoreCache *lru.Cache
+
+	// caches the blocks themselves -> can be scaled by using a redis/memcache etc distributed cache.
+	// multihash -> block
+	blockCache *lru.Cache
+}
+
+func (d *DAGStore) AllShardsReadBlockstore(shardSelector ShardSelectorF, maxCacheSize int, maxBlocks int) (blockstore.Blockstore, error) {
+	// instantiate the blockstore cache
+	bslru, err := lru.NewWithEvict(maxCacheSize, func(_ interface{}, val interface{}) {
+		// ensure we close the blockstore for a shard when it's evicted from the cache so dagstore can gc it.
+		abs := val.(*accessorWithBlockstore)
+		abs.sa.Close()
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create lru cache for read only blockstores")
+	}
+
+	// instantiate the block cache
+	blkLru, err := lru.New(maxBlocks)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create lru cache for blocks: %w", err)
+	}
+
+	return &AllShardsReadBlockstore{
+		d:               d,
+		shardSelectF:    shardSelector,
+		blockstoreCache: bslru,
+		blockCache:      blkLru,
+	}, nil
+}
+
+func (ro *AllShardsReadBlockstore) Get(ctx context.Context, c cid.Cid) (b blocks.Block, finalErr error) {
+	logbs.Debugw("bitswap Get called", "cid", c)
+	defer func() {
+		if finalErr != nil {
+			logbs.Debugw("bitswap Get: got error", "cid", c, "error", finalErr)
+		}
+	}()
+
+	mhash := c.Hash()
+	// do we have the block cached ?
+	if val, ok := ro.blockCache.Get(mhash.String()); ok {
+		logbs.Debugw("bitswap Get: returning from block cache", "cid", c)
+		return val.(blocks.Block), nil
+	}
+
+	// fetch all the shards containing the multihash
+	shards, err := ro.d.ShardsContainingMultihash(ctx, mhash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch shards containing the block: %w", err)
+	}
+	if len(shards) == 0 {
+		return nil, errors.New("no shards contain the requested block")
+	}
+
+	// do we have a cached blockstore for a shard containing the required block ? If yes, serve the block from that blockstore
+	for _, sk := range shards {
+		// a valid cache hit here updates the priority of the shard's blockstore in the LRU cache.
+		val, ok := ro.blockstoreCache.Get(sk)
+		if !ok {
+			continue
+		}
+
+		rbs := val.(*accessorWithBlockstore).bs
+		blk, err := rbs.Get(ctx, c)
+		if err != nil {
+			ro.blockstoreCache.Remove(sk)
+			continue
+		}
```

Comment:
Could you add a comment about why we should remove the blockstore with that shard if there's an error getting it?

Reply:
This is done in the new `IndexBackedBlockstore ` type.